### PR TITLE
Purge room when all users have forgotten it

### LINKED ISSUE #4720: garbage collect forgotten rooms
if everyone on a server has left and forgotten a given room, we should GC it from the DB

### PR DESCRIPTION
**Pull request by pingiun**
Somewhat adresses #4720, based on comment: https://github.com/matrix-org/synapse/issues/4720#issuecomment-671899124

This will only work for newly forgotten rooms, and only rooms that are completely forgotten by all users. This is not a general fix for people who want to remove rooms after some time and after they have been left by everyone.

Signed-off-by: Jelle Besseling <jelle@pingiun.com>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/storage/databases/main/roommember.py [hunk 1]
```diff
@@ -793,6 +793,21 @@ async def is_host_joined(self, room_id: str, host: str) -> bool:
     async def is_host_invited(self, room_id: str, host: str) -> bool:
         return await self._check_host_room_membership(room_id, host, Membership.INVITE)
 
+    async def is_locally_forgotten_room(self, room_id: str) -> bool:
+        sql = """
+            SELECT 1 FROM room_memberships
+            WHERE room_id = ?
+                AND user_id LIKE ?
```
Comment (dklimpel) [inline synapse/storage/databases/main/roommember.py:800]: What are you doing with the `user_id`. Is it required?

There are other examples to get a bool from the database. E.g.:
https://github.com/matrix-org/synapse/blob/00a67f831adc7c01fcced6269cb5790816252493/synapse/storage/databases/main/registration.py#L443-L460

Comment (dklimpel) [inline synapse/storage/databases/main/roommember.py:800]: What are you doing with the `user_id`. Is it required?

There are other examples to get a bool from the database. E.g.:
https://github.com/matrix-org/synapse/blob/00a67f831adc7c01fcced6269cb5790816252493/synapse/storage/databases/main/registration.py#L443-L460

Comment (pingiun) [reply synapse/storage/databases/main/roommember.py:800]: My thinking with this is that remote rooms should be purged locally if all the local users have forgotten about it. I'm not 100% sure if purge works this way because I only tested it with local rooms anyways.

Comment (pingiun) [reply synapse/storage/databases/main/roommember.py:800]: In other words, it should not be required that remote users forget about a remote room, only local users are counted by this LIKE clause

File: synapse/storage/databases/main/roommember.py [hunk 2]
```diff
@@ -793,6 +793,21 @@ async def is_host_joined(self, room_id: str, host: str) -> bool:
...
+                AND forgotten != 1
+            LIMIT 1
```
Comment (erikjohnston) [inline synapse/storage/databases/main/roommember.py:802]:
```suggestion
            SELECT count(*) FROM local_current_membership
            INNER JOIN room_memberships USING (room_id, event_id)
            WHERE 
                room_id = ?
                AND forgotten = 0;
```

I think ^ is an easier query, and returns the number of users who haven't forgotten the room. This avoids the `LIKE` heuristic, and handles the case where a user leaves a room, forgets it, then rejoins (as `room_memberships` is a table of *all* membership events that have happened, not just the current ones).

File: synapse/rest/client/room.py [hunk 3]
```diff
@@ -757,6 +760,14 @@ async def on_POST(
 
         await self.room_member_handler.forget(user=requester.user, room_id=room_id)
 
+        forgotten = await self.store.is_locally_forgotten_room(room_id)
+        if forgotten:
+            run_as_background_process(
+                "purge_room_after_last_forget",
+                self.storage.purge_events.purge_room,
+                room_id,
+            )
```
Comment (erikjohnston) [inline synapse/rest/client/room.py:769]: This is an OK first step, but it does mean that if the server gets restarted before the purge is completed it won't be retried.

One option would be to check if we need to purge the room and if so write that to a DB table, and once purged we can remove it from the table.

That can be a separate PR though I think

File: changelog.d/12170.bugfix [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Purge a room completely when all local users have forgotten it.
+
+Contributed by @pingiun.
```
Comment (erikjohnston) [inline changelog.d/12170.bugfix:3]:
```suggestion
Purge a room completely when all local users have forgotten it. Contributed by @pingiun.
```

This file gets merged into the changelog, which expects entries to be on one line.

**Review by erikjohnston — COMMENTED** [review]
Could you add some tests please? Just a couple to check that we *don't* purge rooms unless everyone has forgotten the room, and one to check that forgetting then rejoining the room works. I'm just very paranoid when it comes to things that delete lots of data :slightly_smiling_face: 

Thanks for this!

File: synapse/storage/databases/main/roommember.py (hunk 2 above)
Comment (pingiun) [reply synapse/storage/databases/main/roommember.py:802]: I think that case is already handled, because it would add a new entry to room_memberships that has forgotten = 0, so my query would return something. But I think this is still good because it avoids the LIKE clause

Comment (erikjohnston) [reply synapse/storage/databases/main/roommember.py]: Oh yes, very true.

**Comment by pingiun** [comment]
I think the failing tests are unrelated to my changes? Not sure

**Comment by erikjohnston** [comment]
```
synapse/storage/databases/main/roommember.py:800:18: W291 trailing whitespace
```

Maybe my suggestion accidentally included a trailing whitespace? Hard to tell on the web IDE

**Comment by erikjohnston** [comment]
Oh, sorry! Yeah, that looks like a flakey test to me, since its just failed on one run and appears unrelated.

**Comment by anoadragon453** [comment]
Hi @pingiun. Are you able to continue working on this? If you need help with writing tests, I'd suggest taking a look at an existing unit test and adapting it.

Awkwardly I'd point you to some existing unit tests for forgetting a room, but it looks like we only have [integration tests for that functionality](https://github.com/matrix-org/sytest/blob/4784b7ed3337962e7938d368d2748953d627d5bf/tests/30rooms/31forget.pl)!

I'd instead suggest creating a new testfile in [tests/rest/client/](https://github.com/matrix-org/synapse/tree/develop/tests/rest/client) named `test_forget.py` or similar and crib off an existing test file for the usual unit test structure. We're also open to providing help over in our dev channel if you need any :)

https://matrix.to/#/#synapse-dev:matrix.org

**Comment by anoadragon453** [comment]
Closing this PR as it seems to have been abandoned. If you'd like to continue working on this, feel free to re-open / create a new PR. Thanks!
